fix(backend): ensure that files of released entries are accessible (#4606)
loculus-project/loculus: backend/docs/file_sharing.md

The preprocessing pipeline has extra permissions and can also upload new files on behalf of the submitting group.

## Accessing a processed file

To access a processed file, the `/get/{accession}/{version}/{fileCategory}/{fileName}` endpoint can be called. If the sequence entry is released, the file is public and can be accessed without authentication. If the sequence entry is not released, it can only be accessed by authenticated and authorized users.

If the file exists and may be accessed by the user, the endpoint will return a `307 redirect`-response, redirecting a user to a pre-signed URL to receive the file.

## S3 file visibility

By default, the files on S3 are private. When a sequence entry is released, the associated files are made public in S3 as well. The bucket is configured to allow public access for objects which are tagged with `public=true`,
so to make a file public, the backend sets this tag on the object in S3.

> ### Releasing and pipeline upgrades

Author: This is an open issue: #4056. (You can find a list of open issues here: #4221)

There are currently two mechanisms to handle files of new pipeline versions:

> - When files are uploaded for entries that are already public, the files are [published right away](https://github.com/loculus-project/loculus/pull/4515).

Author: We currently have different options in the config `fileSharing.outputFileUrlType` regarding which file URLs are exposed (i.e., sent to LAPIS and shown in the downloadable files):

- If `s3`, then the URLs can look like `hel1.your-objectstorage.com/loculus/files/78e045cc-ce4b-4bd6-be47-f85306e8aae4`. In this case, users will directly access files on S3 without any authentication token/signature. To use this, the file must be public in S3, so we need to tag the file with `public=true` on S3.
- If `backend`, then the URLs can look like `backend.pathoplexus.org/seq/files/get/PP12345/1/rawData/r1.fastq.gz`. The response will be a redirect to a pre-signed S3 URL like `hel1.your-objectstorage.com/loculus/files/78e045cc-ce4b-4bd6-be47-f85306e8aae4?<signature parameters>`. Because the URL is pre-signed, the file does not need to be actually public on S3. With this PR, a file is public when the associated sequence entry is public, so there is no dedicated publishing step.
- If `website`, then the URLs can look like `pathoplexus.org/seq/PP12345.1/rawData/r1.fastq.gz`. This calls the backend endpoint to get the URL to S3 and then redirects the users.

(As just also mentiond in https://github.com/loculus-project/loculus/pull/4606#discussion_r2244662595), we can reconsider in a while, whether we actually need the `s3` option or would like to remove it.

> - When files are uploaded for entries that are already public, the files are [published right away](https://github.com/loculus-project/loculus/pull/4515).

Reviewer: Ah do we actually have an issue to track if we should remove the s3 option?

> - When files are uploaded for entries that are already public, the files are [published right away](https://github.com/loculus-project/loculus/pull/4515).

Author: Created the issue: #4805

> - When files are uploaded for entries that are not yet published, but are also uploaded with a pipeline version greater than the current one, the backend will [_also_ publish these already](https://github.com/loculus-project/loculus/pull/4581). This is to prevent the files from being private once the current pipeline version is bumped. Why not publish these files during the pipeline version bump? The database is not locked during the bump, and so a file that is uploaded with the n+1 pipeline version could be missed.

Reviewer: Ah I think this sentence might actually be badly worded. https://github.com/loculus-project/loculus/pull/4581 publishes (=makes public by adding a public flag) all files from the current and future pipelines when a file from the current pipeline is approved. The current wording makes it sound like future files are published even if no version has been approved.

> - When files are uploaded for entries that are not yet published, but are also uploaded with a pipeline version greater than the current one, the backend will [_also_ publish these already](https://github.com/loculus-project/loculus/pull/4581). This is to prevent the files from being private once the current pipeline version is bumped. Why not publish these files during the pipeline version bump? The database is not locked during the bump, and so a file that is uploaded with the n+1 pipeline version could be missed.

Author: I changed it to "will publish these when the entry is approved".

## The files table

> The database has a table to keep track of the files.

Reviewer 2: ```suggestion
The database keeps track of files in the `files` table.
```
